crates/chain-config/src/config/codec/decoder.rs

```diff
@@ -0,0 +1,191 @@
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+enum DataSource {
+    #[cfg(feature = "parquet")]
+    Parquet { files: crate::ParquetFiles },
+    InMemory {
+        state: StateConfig,
+        group_size: usize,
+    },
+}
+
+#[derive(Clone, Debug)]
+pub struct Decoder {
+    data_source: DataSource,
+}
+
+impl Decoder {
+    #[cfg(feature = "std")]
+    pub fn json(
+        path: impl AsRef<std::path::Path>,
+        group_size: usize,
+    ) -> anyhow::Result<Self> {
+        let mut file = std::fs::File::open(path)?;
+
+        let state = serde_json::from_reader(&mut file)?;
+
+        Ok(Self::in_memory(state, group_size))
+    }
+
+    pub fn in_memory(state: StateConfig, group_size: usize) -> Self {
+        Self {
+            data_source: DataSource::InMemory { state, group_size },
+        }
+    }
+
+    #[cfg(feature = "parquet")]
+    pub fn parquet(files: crate::ParquetFiles) -> Self {
+        Self {
+            data_source: DataSource::Parquet { files },
+        }
+    }
+
+    #[cfg(feature = "std")]
+    pub fn for_snapshot(
+        snapshot_metadata: crate::config::SnapshotMetadata,
+        default_group_size: usize,
+    ) -> anyhow::Result<Self> {
+        use crate::StateEncoding;
+
+        let decoder = match snapshot_metadata.take_state_encoding() {
+            StateEncoding::Json { filepath } => Self::json(filepath, default_group_size)?,
+            #[cfg(feature = "parquet")]
+            StateEncoding::Parquet { filepaths, .. } => Self::parquet(filepaths),
+        };
+        Ok(decoder)
+    }
+
+    pub fn coins(&self) -> anyhow::Result<IntoIter<CoinConfig>> {
+        self.create_iterator(
+            |state| &state.coins,
+            #[cfg(feature = "parquet")]
+            |files| &files.coins,
+        )
+    }
+
+    pub fn messages(&self) -> anyhow::Result<IntoIter<MessageConfig>> {
+        self.create_iterator(
+            |state| &state.messages,
+            #[cfg(feature = "parquet")]
+            |files| &files.messages,
+        )
+    }
+
+    pub fn contracts(&self) -> anyhow::Result<IntoIter<ContractConfig>> {
+        self.create_iterator(
+            |state| &state.contracts,
+            #[cfg(feature = "parquet")]
+            |files| &files.contracts,
+        )
+    }
+
+    pub fn contract_state(&self) -> anyhow::Result<IntoIter<ContractStateConfig>> {
+        self.create_iterator(
+            |state| &state.contract_state,
+            #[cfg(feature = "parquet")]
+            |files| &files.contract_state,
+        )
+    }
+
+    pub fn contract_balance(&self) -> anyhow::Result<IntoIter<ContractBalance>> {
+        self.create_iterator(
+            |state| &state.contract_balance,
+            #[cfg(feature = "parquet")]
+            |files| &files.contract_balance,
+        )
+    }
+
+    fn create_iterator<T: Clone>(
+        &self,
+        extractor: impl FnOnce(&StateConfig) -> &Vec<T>,
+        #[cfg(feature = "parquet")] file_picker: impl FnOnce(
```

Comment:
I think this should just be a different function for `feature = "parquet"`. Having a different function signature depending on the features sounds like a pitfall :hole:

Reply:
I guess it's not public so it's not the worst. It just feels a little clever. I'll downgrade this to a "Nit"